src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs

```diff
@@ -191,6 +301,829 @@ public static void Main()
 O brave new world...
 ";
             CompileAndVerify(source, expectedOutput: expected);
+
+            var comp = CreateCompilation([source, RuntimeAsyncAwaitHelpers], targetFramework: TargetFramework.Net90);
+            comp.Assembly.SetOverrideRuntimeSupportsAsyncMethods();
+
+            var verifier = CompileAndVerify(comp, verify: Verification.Fails with
+            {
+                ILVerifyMessage = "[F]: Unexpected type on the stack. { Offset = 0x2e, Found = ref 'string', Expected = ref '[System.Runtime]System.Threading.Tasks.Task`1<string>' }",
+            }, symbolValidator: verify);
+            verifier.VerifyDiagnostics();
+
+            verifier.VerifyIL("Test.F", """
+                {
+                  // Code size       47 (0x2f)
+                  .maxstack  3
+                  IL_0000:  call       "System.Threading.Tasks.TaskFactory System.Threading.Tasks.Task.Factory.get"
+                  IL_0005:  ldsfld     "System.Func<string> Test.<>c.<>9__0_0"
+                  IL_000a:  dup
+                  IL_000b:  brtrue.s   IL_0024
+                  IL_000d:  pop
+                  IL_000e:  ldsfld     "Test.<>c Test.<>c.<>9"
+                  IL_0013:  ldftn      "string Test.<>c.<F>b__0_0()"
+                  IL_0019:  newobj     "System.Func<string>..ctor(object, nint)"
+                  IL_001e:  dup
+                  IL_001f:  stsfld     "System.Func<string> Test.<>c.<>9__0_0"
+                  IL_0024:  callvirt   "System.Threading.Tasks.Task<string> System.Threading.Tasks.TaskFactory.StartNew<string>(System.Func<string>)"
+                  IL_0029:  call       "string System.Runtime.CompilerServices.RuntimeHelpers.Await<string>(System.Threading.Tasks.Task<string>)"
+                  IL_002e:  ret
+                }
+                """);
+
+            void verify(ModuleSymbol module)
+            {
+                var test = module.ContainingAssembly.GetTypeByMetadataName("Test");
+                var f = test.GetMethod("F");
+                Assert.Equal(MethodImplOptionsAsync, f.ImplementationAttributes);
+                AssertEx.Equal(["<>c"], test.GetTypeMembers().SelectAsArray(t => t.Name));
+            }
+        }
+
+        [Fact]
+        public void GenericValueTaskReturningAsync()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+
+class Test
+{
+    public static async ValueTask<string> F()
+    {
+        return await Impl();
+
+        ValueTask<string> Impl() => ValueTask.FromResult(""O brave new world..."");
+    }
+
+    public static async Task Main()
+    {
+        string s = await F();
+        Console.WriteLine(s);
+    }
+}";
+            //var expected = @"O brave new world...";
```

Comment:
It feels like these expected could be uncommented and wrapped in an `IncludeExpectedOutput()` helper similar to tests for DIMs and other runtime-dependent features. The helper would always return null for now. But once we can get a runtime which runs these, the helper could be adjusted to simply check `ExecutionConditionUtil.IsMonoOrCoreClr`, and all the tests start expecting the right thing when the test runtime supports it. Not something you need to do, but thought maybe the idea would be helpful.